src/main/scala/analysis/MemoryModelMap.scala

```diff
@@ -254,20 +522,125 @@ class MemoryModelMap {
       logRegion(range, region)
     }
   }
+
+  def mergeRegions(regions: Set[MemoryRegion]): MemoryRegion = {
+    // assert regions are of the same type
+    regions.foreach(uf.makeSet)
+    regions.foreach(uf.union(regions.head, _))
+    uf.find(regions.head)
+  }
+
+  private def returnRegion(region: MemoryRegion): MemoryRegion = {
+    uf.find(region)
+  }
+
+  private def returnRegion(region: StackRegion): StackRegion = {
+    uf.find(region.asInstanceOf[MemoryRegion]).asInstanceOf[StackRegion]
+  }
+
+  private def returnRegion(region: DataRegion): DataRegion = {
+    uf.find(region.asInstanceOf[MemoryRegion]).asInstanceOf[DataRegion]
+  }
+
+  private def returnRegion(region: HeapRegion): HeapRegion = {
+    uf.find(region.asInstanceOf[MemoryRegion]).asInstanceOf[HeapRegion]
+  }
+
+  def getHeap(directCall: DirectCall): HeapRegion = {
+    require(directCall.target.name == "malloc", "Should be a malloc call")
+    heapCalls(directCall)
+  }
+
+  def getStack(allocationSite: CFGPosition): Set[StackRegion] = {
+    stackAllocationSites.getOrElse(allocationSite, Set.empty).map(returnRegion)
+  }
+
+  def getData(cfgPosition: CFGPosition): Set[DataRegion] = {
+    cfgPositionToDataRegion.getOrElse(cfgPosition, Set.empty).map(returnRegion)
+  }
+
+  def addMergeRegions(regions: Set[MemoryRegion], name: String): Unit = {
+    mergedRegions(regions) = name
+  }
+
+  def getMergedName(regions: Set[MemoryRegion]): String = {
+    mergedRegions(regions)
+  }
 }
 
 trait MemoryRegion {
   val regionIdentifier: String
+  val start: BigInt
+  val subAccesses: mutable.Set[BigInt] = mutable.Set()
 }
 
-case class StackRegion(override val regionIdentifier: String, start: BitVecLiteral, parent: Procedure) extends MemoryRegion {
-  override def toString: String = s"Stack($regionIdentifier, $start, ${parent.name})"
+case class StackRegion(override val regionIdentifier: String, start: BigInt, parent: Procedure) extends MemoryRegion {
+  override def toString: String = s"Stack($regionIdentifier, $start, ${parent.name}, $subAccesses)"
 }
 
-case class HeapRegion(override val regionIdentifier: String, size: BitVecLiteral, parent: Procedure) extends MemoryRegion {
+case class HeapRegion(override val regionIdentifier: String, start: BigInt, size: BigInt, parent: Procedure) extends MemoryRegion {
   override def toString: String = s"Heap($regionIdentifier, $size)"
 }
 
-case class DataRegion(override val regionIdentifier: String, start: BitVecLiteral) extends MemoryRegion {
-  override def toString: String = s"Data($regionIdentifier, $start)"
+case class DataRegion(override val regionIdentifier: String, start: BigInt, size: BigInt) extends MemoryRegion {
+  override def toString: String = s"Data($regionIdentifier, $start, $size, ($relfContent))"
+  def end: BigInt = start + size - 1
+  val relfContent: mutable.Set[String] = mutable.Set[String]()
```

Comment:
It's bad practice to put mutable parameters like this in a case class (or in this case, a trait that case classes extend).